articles/152_roslaunch_multi_machine_launch.md

```diff
@@ -0,0 +1,310 @@
+
+## Purpose
+
+Allow a system of ROS nodes to be launched on a hardware architecture that is
+spread across multiple networked computers and facilitate introspection and
+management of the system from a single machine.
+
+## Justification
+
+Nodes can need to run on different hosts for a variety of reasons.  Some possible
+use cases:
+
+- A large robot could with hosts located physically near the hardware they are
+controlling such as cameras or other sensors
+- A robot with hosts with different architectures in order to use specialized
+processing hardware
+- A robot with a cluster of machines that do distributed processing of data
+- A network of multiple virtual hosts for testing purposes
+- A swarm of independent drones that can cooperate but do not require
+communication with each other
+- A client computer is used to launch and monitor a system on a remote host but
+is not required for the system to operate
+
+## Capabilities
+
+In order to meet the above use cases, a launch system needs to have a number of
+capabilities, including:
+
+- Connecting to a remote host and running nodes on it
+- Pushing configuration parameters for nodes to remote hosts
+- Monitoring the status and tracking the lifecycle of nodes
+- Recovering from failures by optionally restarting nodes
+- Gracefully shutting down nodes on every host
+
+Most of these are just extensions of the design goals for the single-machine
+version of roslaunch for ROS2, but in addition to extending those goals to
+remote machines, we would also like to consider a few more advanced features,
+including:
+
+- Load balancing nodes on distributed networks
+- Command line tools for managing and monitoring systems across machines
+- Mechanisms for locating files and executables across machines
+- Sharing and synchronizing files across machines
+
+## Considerations
+
+There are some outstanding issues that may complicate things:
+
+- How to group nodes/participants/processes is somewhat of an open issue with potential
+   implications for this part of ROS2.
+    - https://github.com/ros2/design/pull/250/files/8ccaac3d60d7a0ded50934ba6416550f8d2af332?short_path=dd776c0#diff-dd776c070ecf252bc4dcc4b86a97c888
+    - The number of domain participants is limited per vendor (Connext is 120 per domain).
+- No `rosmaster` means there is no central mechanism for controlling modes or
+  distributing parameters
+- Machines may be running different operating systems
+- If we intend to do any kind of load balancing, certain types of resources may
+    need to be transferred to other machines.
+     - Calibration data, map files, training data, etc.
+     - Need to keep track of which machine has the most recent version of such
+        resources
+- Security: we'll need to manage credentials across numerous machines both for SSH
+    and secure DDS.
+
+
+## Proposed Multi-Machine Launch Command Line Interface
+
+The multi-machine launching interface is controlled through the `launcher`
```

Comment:
@mlanting meta: I wonder if `launcher` is the right naming, as opposed to `host`, `machine` or even `system`.

Reply:
It was the best thing I could think of at the time -- I am also not a big fan of it and an open to suggestions. :-)

I think it would make sense to provide the same interface for both single-machine and multi-machine launching and also make it obvious from the verb what the command does.  `system` might work for that.

Reply:
`launcher` is just a new command we decided to propose since the existing `launch` command doesn't have any verbs. Running something like `ros2 launcher launch` would do the same thing as `ros2 launch`, either by calling the existing launch, or replacing it (in which case `ros2 launch` could just be an alias).

This is basically just a workaround for the fact that the existing launch command doesn't have verbs and we weren't sure it'd be a good idea to try to introduce verbs to a verbless command.